File: extensions/shared/library/src/main/java/app/morphe/extension/shared/addons/AddonModule.java
package app.morphe.extension.shared.addons;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.Collections;
import java.util.Map;
import java.util.Set;

/**
 * Interface for runtime addon modules loaded in the patched app.
 *
 * <p>Modules are discovered from DEX files placed in the app's addon directory.
 * Each DEX must contain a class implementing this interface, declared in its
 * {@code addon.properties} manifest.</p>
 *
 * <p>Example module:</p>
 * <pre>
 * public class SkipArtistsModule implements AddonModule {
 *     public String getId() { return "skip-artists"; }
 *     public String getName() { return "Skip Artists"; }
 *     public String getDescription() { return "Skips specific artists during playback"; }
 *     public String getVersion() { return "1.0.0"; }
 *
 *     public void onEnable(Map&lt;String, String&gt; config) {
 *         String artists = config.get("artists");
 *         // Register playback hooks...
 *     }
 * }
 * </pre>
 */
public interface AddonModule {

    /** Unique identifier for this module. */
    @NonNull
    String getId();

    /** Human-readable display name. */
    @NonNull
    String getName();

    /** Short description of what this module does. */
    @NonNull
    String getDescription();

    /** Module version string. */
    @NonNull
    String getVersion();

    /**
     * Package names this module is compatible with.
     * Return empty set for universal compatibility.
     */
    @NonNull
    default Set<String> getCompatibleApps() {
        return Collections.emptySet();
    }

    /**
     * Called when the module is enabled. Initialize hooks and state here.
     *
     * @param config User-defined key-value configuration for this module.
     */
    default void onEnable(@NonNull Map<String, String> config) {}

    /**
     * Called when the module is disabled. Clean up hooks and state here.
     */
    default void onDisable() {}

    /**
     * Called when the module's configuration changes while it is enabled.
     *
     * @param config Updated configuration.
     */
    default void onConfigChanged(@NonNull Map<String, String> config) {}
}

File: patches/src/main/kotlin/app/morphe/patches/all/misc/addons/AddonModule.kt
package app.morphe.patches.all.misc.addons

/**
 * Interface for custom addon modules that can be loaded alongside patches.
 *
 * Addon modules provide a way to extend the patching process with custom logic,
 * such as skipping specific artists, filtering content, or applying conditional modifications.
 *
 * Example usage:
 * ```kotlin
 * class SkipArtistModule : AddonModule {
 *     override val id = "skip-artists"
 *     override val name = "Skip Artists"
 *     override val description = "Skips playback for specific artists"
 *     override val version = "1.0.0"
 *
 *     override fun onLoad(config: Map<String, Any>) {
 *         // Initialize with user-provided artist list
 *     }
 *
 *     override fun onExecute(context: AddonContext) {
 *         // Apply the filtering logic
 *     }
 * }
 * ```
 */
interface AddonModule {
    /** Unique identifier for this module. */
    val id: String

    /** Human-readable name. */
    val name: String

    /** Short description of what this module does. */
    val description: String

    /** Module version string. */
    val version: String

    /**
     * Compatible target apps this module can work with.
     * Empty set means compatible with all apps.
     */
    val compatibleApps: Set<String>
        get() = emptySet()

    /**
     * Called when the module is loaded. Use this to initialize state
     * and read configuration values.
     *
     * @param config Key-value configuration passed by the user.
     */
    fun onLoad(config: Map<String, Any>) {}

    /**
     * Called during patch execution. This is where the module applies
     * its custom logic.
     *
     * @param context Provides access to patching context and utilities.
     */
    fun onExecute(context: AddonContext) {}

    /**
     * Called after all patches have been applied. Use this for cleanup
     * or final modifications.
     */
    fun onFinalize() {}
}
